[LT] Support diagonal op

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75230

Summary:
Op diagonal is a view op which we can't code-gen yet. Therefore, support
it by making hand-written IR construction and lowering.

Test Plan:
./build/bin/test_lazy --gtest_filter=LazyOpsTest.TestDiagonal*

Differential Revision: [D35378316](https://our.internmc.facebook.com/intern/diff/D35378316)

Commit:
[LT] Support diagonal op

Summary:
Op diagonal is a view op which we can't code-gen yet. Therefore, support
it by making hand-written IR construction and lowering.

Test Plan:
./build/bin/test_lazy --gtest_filter=LazyOpsTest.TestDiagonal

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35378316).

wconstab reviewed (approved):
lgtm. assuming, it is covered by our existing tests?

alanwaketan commented:
> LazyOpsTest.TestDiagonal

Yea, we have LazyOpsTest.TestDiagonal*. Thanks for approving this change.

Commit:
Update on "[LT] Support diagonal op"


Summary:
Op diagonal is a view op which we can't code-gen yet. Therefore, support
it by making hand-written IR construction and lowering.

Test Plan:
./build/bin/test_lazy --gtest_filter=LazyOpsTest.TestDiagonal

Differential Revision: [D35378316](https://our.internmc.facebook.com/intern/diff/D35378316)

[ghstack-poisoned]

Commit:
Update on "[LT] Support diagonal op"


Summary:
Op diagonal is a view op which we can't code-gen yet. Therefore, support
it by making hand-written IR construction and lowering.

Test Plan:
./build/bin/test_lazy --gtest_filter=LazyOpsTest.TestDiagonal

Differential Revision: [D35378316](https://our.internmc.facebook.com/intern/diff/D35378316)

[ghstack-poisoned]

alanwaketan commented on torch/csrc/lazy/ts_backend/ts_node_lowering.cpp:
```diff
@@ -346,6 +356,37 @@ class TSNodeLowering : public TSNodeLoweringInterface {
+
+  // FIXME(alanwaketan): One day we should code-gen all view ops, or at
+  // least move the lowering to the IR nodes.
+  TSOpVector LowerDiagonalViewUpdate(const DiagonalViewUpdate* node) {
```
@Krovatkin What do you think about this approach to update the original tensor/view's values? Do you think it can be generalized, such as your `expand` case.?

Krovatkin replied:
yeah, I think we can use this approach in general

alanwaketan replied:
@bdhirsh Thought you might be interested in these.

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35378316).

Commit:
Update on "[LT] Support diagonal op"


Summary:
Op diagonal is a view op which we can't code-gen yet. Therefore, support
it by making hand-written IR construction and lowering.

Test Plan:
./build/bin/test_lazy --gtest_filter=LazyOpsTest.TestDiagonal*

Differential Revision: [D35378316](https://our.internmc.facebook.com/intern/diff/D35378316)

[ghstack-poisoned]

Commit:
Update on "[LT] Support diagonal op"


Summary:
Op diagonal is a view op which we can't code-gen yet. Therefore, support
it by making hand-written IR construction and lowering.

Test Plan:
./build/bin/test_lazy --gtest_filter=LazyOpsTest.TestDiagonal*

Differential Revision: [D35378316](https://our.internmc.facebook.com/intern/diff/D35378316)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35378316).
